### DISCUSSION TITLE
Use a more dynamic memory model so other microcontrolles can be supported

### DISCUSSION THREAD

**Discussion by beachviking** [pschatzmann/SIDPlayer#8] [Ideas]
I found a SID player for the STM32 microcontrollers here:
https://github.com/Bakisha/STM32-SID-PLAYER

This seems to be able to use less memory and provides for some kind of dynamic memory mapping?

Could serve as a template for an improved SID player, but the code is not structured well and it will be some work to port the relevant sections.

**Comment by pschatzmann** [comment]
Not sure if this really worth the effort. 
After all a AudioKit has 4MB PSRAM that we can use...

**Reply by beachviking** [reply]
Agreed. Just a potential roadmap item, if anyone feels compelled or have a special need in their projects. :)

**Reply by pschatzmann** [reply]
I moved the MAX_DATA_LEN to SIDConfig.h. 
When playing around with some values I could make the commando example song play with  25000

**Reply by beachviking** [reply]
I see.  

In looking at https://github.com/Bakisha/STM32-SID-PLAYER/tree/master/InternalFlash, I suspect a mechanism could be created that will allocate memory based on the expected Sid load address coming from the PSID data + the length of the PSID data? May not work for all PSIDs, but maybe for most of them?

**Reply by beachviking** [reply]
Ref.

https://github.com/Bakisha/STM32-SID-PLAYER/blob/master/InternalFlash/STM32_SID_player_internal/10_Loader.ino

**Reply by pschatzmann** [reply]
I would leave this up to the user. 
Otherwise you will constantly get issues from other people that will complain if it does not work.

**Reply by pschatzmann** [reply]
But I think the following option would be interesting: When the song is in progmem we would not allocate and copy the data, but just use it directly from progmem.  That would keep the RAM use to a minimum.
This approach should also work when the data is loaded from a file via RAM and void avoid the need for double allocations

This way it would also work on something as small as a Arduino Nano

**Reply by pschatzmann** [reply]
I tried to provide an experimental implementation that only needs around 2K of RAM.
See https://github.com/pschatzmann/SIDPlayer/blob/main/src/sid-api/sidmemory.h

**Reply by beachviking** [reply]
Interesting!  If it works, that would be cool!

**Reply by pschatzmann** [reply]
I did some rough testing and It's playing music.

Out of curiosity I was wondering what this would give with an Nano:
```
Error compiling for board Arduino Nano.
Sketch uses 62228 bytes (202%) of program storage space. Maximum is 30720 bytes.
Global variables use 31401 bytes (1533%) of dynamic memory, leaving -29353 bytes for local variables. Maximum is 2048 bytes.
```

**Comment by beachviking** [comment]
Hope to test this later today.  Very cool.

I did come across another project which is using a Pi Pico and is using a SID implementation from RockBox which looks code wise quite nice.  Seems to also have been optimized for low memory devices as well. I am putting the link here as a reference.  

https://github.com/henrikenblom/SIDPod

**Comment by pschatzmann** [comment]
https://github.com/henrikenblom/SIDPod/blob/master/src/audio/sid.c: unsigned char memory[65536];
I guess this is also based on cSID light
